Problems parsing SwitchExpr with Unary/BinaryOperators

Hi,

I have problems parsing SwitchExprs with UnaryOperators. These problems occur in the not released snapshot version. The released version 3.25.10 works correctly. So I think these problems were added, while implementing the Guards into SwitchExpr. 

Here are test cases and the corresponding output:

Test cases:
```Java
    @Test
    void testSwitchExprUnaryMinus() {
        parseStatement("int j = switch (i) {\n" +
                       "            case 0 -> 0;\n" +
                       "            default -> -1;\n" +
                       "        };");
    }

    @Test
    void testSwitchExprUnaryNot() {
        parseStatement("boolean j = switch (i) {\n" +
                       "            case 0 -> true;\n" +
                       "            default -> !false;\n" +
                       "        };");
    }
```

Output testSwitchExprUnaryMinus():
```
[ERROR] Failures: 
[ERROR]   SwitchExprTest.testSwitchExprUnaryMinus:293 [(line 3,col 21) Parse error. Found "-", expected one of  "(" "++" "--" "@" "\"\"\"" "boolean" "byte" "char" "double" "enum" "exports" "false" "float" "int" "long" "module" "new" "null" "open" "opens" "permits" "provides" "record" "requires" "sealed" "short" "strictfp" "super" "this" "throw" "to" "transitive" "true" "uses" "void" "when" "with" "yield" "{" <CHARACTER_LITERAL> <FLOATING_POINT_LITERAL> <IDENTIFIER> <INTEGER_LITERAL> <LONG_LITERAL> <STRING_LITERAL>
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:14249)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:14094)
  com.github.javaparser.GeneratedJavaParser.SwitchEntry(GeneratedJavaParser.java:6562)
[...]
```

Output testSwitchExprUnaryNot():
```
[ERROR]   SwitchExprTest.testSwitchExprUnaryNot:301 [(line 3,col 21) Parse error. Found "!", expected one of  "(" "++" "--" "@" "\"\"\"" "boolean" "byte" "char" "double" "enum" "exports" "false" "float" "int" "long" "module" "new" "null" "open" "opens" "permits" "provides" "record" "requires" "sealed" "short" "strictfp" "super" "this" "throw" "to" "transitive" "true" "uses" "void" "when" "with" "yield" "{" <CHARACTER_LITERAL> <FLOATING_POINT_LITERAL> <IDENTIFIER> <INTEGER_LITERAL> <LONG_LITERAL> <STRING_LITERAL>
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:14249)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:14094)
  com.github.javaparser.GeneratedJavaParser.SwitchEntry(GeneratedJavaParser.java:6562)
[...]
```

It seems to be a parser problem. In both cases the UnaryOperator after the `->` symbol is the problem. It's also not a problem of the `default` case, but also occurs in other cases too.

Could someone please look into it?
I am happy to help, but I am not familiar with JavaCC.


Best regards
Kim

Apparently it is not only unary operators in a SwitchEntry, but also binary operators that cause parser errors. 

An additional test follows:

```Java
    @Test
    void testSwitchExprWithBinaryExpr() {
        parseStatement("int i = switch (1) {\n" +
                       "    case 1 -> 1;\n" +
                       "    case 2, 3 -> 1 + 2;\n" +
                       "    default -> 1;\n" +
                       "};");
    }
```

Output:
```
[ERROR]   SwitchExprTest.testSwitchExprWithBinaryExpr:309 [(line 3,col 18) Parse error. Found "+", expected one of  "%=" "&=" "*=" "++" "+=" "--" "-=" "/=" ";" "<<=" "=" ">>=" ">>>=" "^=" "|="
Problem stacktrace : 
  com.github.javaparser.GeneratedJavaParser.generateParseException(GeneratedJavaParser.java:14249)
  com.github.javaparser.GeneratedJavaParser.jj_consume_token(GeneratedJavaParser.java:14094)
  com.github.javaparser.GeneratedJavaParser.StatementExpression(GeneratedJavaParser.java:6228)
  com.github.javaparser.GeneratedJavaParser.SwitchEntry(GeneratedJavaParser.java:6540)
[...]
```

Best regards

--- a/javaparser-core-testing/src/test/java/com/github/javaparser/ast/expr/SwitchExprTest.java
+++ b/javaparser-core-testing/src/test/java/com/github/javaparser/ast/expr/SwitchExprTest.java
@@ -22,6 +22,7 @@
 package com.github.javaparser.ast.expr;
 
 import com.github.javaparser.ast.NodeList;
+import com.github.javaparser.ast.body.VariableDeclarator;
 import com.github.javaparser.ast.stmt.BlockStmt;
 import com.github.javaparser.ast.stmt.Statement;
 import com.github.javaparser.ast.stmt.SwitchEntry;
@@ -236,7 +237,7 @@ class SwitchExprTest {
     @Test
     void switchPatternWithGuard() {
         SwitchExpr expr = parseExpression("switch (value) {\n" +
-                "    case Box b when b.nonEmpty() -> System.out.println(b);\n" +
+                "    case Box b when b.nonEmpty() -> b.get() + 12;\n" +
                 "}").asSwitchExpr();
 
         assertEquals(1, expr.getEntries().size());
@@ -252,6 +253,8 @@ class SwitchExprTest {
 
         assertEquals("b", label.getNameAsString());
         assertEquals("Box", label.getTypeAsString());
+
+        assertEquals("b.get() + 12;", entry.getStatements().get(0).toString());
     }
 
     @Test
@@ -319,5 +322,98 @@ class SwitchExprTest {
         TypePatternExpr integerPattern = boxPattern.getPatternList().get(0).asTypePatternExpr();
         assertEquals("Integer", integerPattern.getTypeAsString());
         assertEquals("i", integerPattern.getNameAsString());
+    }
+
+    /**
+     * Credit to @Kimmmey in https://github.com/javaparser/javaparser/issues/4440 for the
+     * example code.
+     */
+    @Test
+    void testSwitchExprUnaryMinus() {
+        Statement stmt = parseStatement("int i = switch (x) {\n" +
+                "    case 0 -> 0;\n" +
+                "    default -> -1;\n" +
+                "};");
+
+        VariableDeclarator declarator = (VariableDeclarator) stmt.getChildNodes().get(0).getChildNodes().get(0);
+        SwitchExpr switchExpr = declarator.getInitializer().get().asSwitchExpr();
+
+        assertEquals("0", switchExpr.getEntry(0).getLabels().get(0).toString());
+        assertEquals("0;", switchExpr.getEntry(0).getStatements().get(0).toString());
+
+        assertTrue(switchExpr.getEntry(1).getLabels().isEmpty());
+        assertTrue(switchExpr.getEntry(1).isDefault());
+        assertEquals("-1;", switchExpr.getEntry(1).getStatements().get(0).toString());
+    }
+
+    /**
+     * Credit to @Kimmmey in https://github.com/javaparser/javaparser/issues/4440 for the
+     * example code.
+     */
+    @Test
+    void testSwitchExprUnaryNot() {
+        Statement stmt = parseStatement("boolean b = switch (x) {\n" +
+                "    case 0 -> true;\n" +
+                "    default -> !false;\n" +
+                "};");
+
+        VariableDeclarator declarator = (VariableDeclarator) stmt.getChildNodes().get(0).getChildNodes().get(0);
+        SwitchExpr switchExpr = declarator.getInitializer().get().asSwitchExpr();
+
+        assertEquals("0", switchExpr.getEntry(0).getLabels().get(0).toString());
+        assertEquals("true;", switchExpr.getEntry(0).getStatements().get(0).toString());
+
+        assertTrue(switchExpr.getEntry(1).getLabels().isEmpty());
+        assertTrue(switchExpr.getEntry(1).isDefault());
+        assertEquals("!false;", switchExpr.getEntry(1).getStatements().get(0).toString());
+    }
+
+    /**
+     * Credit to @Kimmmey in https://github.com/javaparser/javaparser/issues/4440 for the
+     * example code.
+     */
+    @Test
+    void testSwitchExprWithBinaryExpr() {
+        Statement stmt = parseStatement("int i = switch (x) {\n" +
+                "    case 1 -> 1;\n" +
+                "    case 2, 3 -> 1 + 2;\n" +
+                "    default -> 1;\n" +
+                "};");
+
+        VariableDeclarator declarator = (VariableDeclarator) stmt.getChildNodes().get(0).getChildNodes().get(0);
+        SwitchExpr switchExpr = declarator.getInitializer().get().asSwitchExpr();
+
+        assertEquals("1", switchExpr.getEntry(0).getLabels().get(0).toString());
+        assertEquals("1;", switchExpr.getEntry(0).getStatements().get(0).toString());
+
+        assertEquals("2", switchExpr.getEntry(1).getLabels().get(0).toString());
+        assertEquals("3", switchExpr.getEntry(1).getLabels().get(1).toString());
+        assertEquals("1 + 2;", switchExpr.getEntry(1).getStatements().get(0).toString());
+
+        assertTrue(switchExpr.getEntry(2).getLabels().isEmpty());
+        assertTrue(switchExpr.getEntry(2).isDefault());
+        assertEquals("1;", switchExpr.getEntry(2).getStatements().get(0).toString());
+
+    }
+
+    @Test
+    void testSwitchExprWithAssignment() {
+        Statement stmt = parseStatement("{\n" +
+                "    int z;\n" +
+                "    int i = switch (x) {\n" +
+                "        case 1 -> z = 1;\n" +
+                "        default -> 1;\n" +
+                "    };\n" +
+                "}");
+
+        VariableDeclarator declarator = (VariableDeclarator) stmt.getChildNodes().get(1).getChildNodes().get(0).getChildNodes().get(0);
+        SwitchExpr switchExpr = declarator.getInitializer().get().asSwitchExpr();
+
+        assertEquals("1", switchExpr.getEntry(0).getLabels().get(0).toString());
+        assertEquals("z = 1;", switchExpr.getEntry(0).getStatements().get(0).toString());
+
+        assertTrue(switchExpr.getEntry(1).getLabels().isEmpty());
+        assertTrue(switchExpr.getEntry(1).isDefault());
+        assertEquals("1;", switchExpr.getEntry(1).getStatements().get(0).toString());
     }
 }

--- a/javaparser-core/src/main/javacc/java.jj
+++ b/javaparser-core/src/main/javacc/java.jj
@@ -4661,6 +4661,7 @@ SwitchEntry SwitchEntry():
     JavaToken begin;
     SwitchEntry ret;
     Statement stmt = null;
+    Expression expr = null;
     boolean isDefault = false;
     Expression guard = null;
 }
@@ -4711,12 +4712,15 @@ SwitchEntry SwitchEntry():
      |
         "->"
         (
-            stmt = StatementExpression()
-            {
-                TokenRange r=range(begin, token());
-                stmts.add(stmt);
-                ret = new SwitchEntry(r, labels, EXPRESSION, stmts, isDefault, guard);
-            }
+            (
+                expr = Expression()
+                {
+                    TokenRange r=range(begin, token());
+                    stmts.add(new ExpressionStmt(expr.getTokenRange().get(), expr));
+                    ret = new SwitchEntry(r, labels, EXPRESSION, stmts, isDefault, guard);
+                }
+                ";"
+            )
          |
             stmt = Block()
             {

